### PR TITLE
GUI trees

### PR DESCRIPTION
**Pull request by AlemSnyder**
This adds trees. The logic should be removed from main, and into a "game engine".

### CODE REVIEW

File: src/Terrain/terrain.cpp [hunk 1]
```diff
@@ -935,7 +909,6 @@ int Terrain::qb_read(const char * path, const std::map<uint32_t, std::pair<const
     file = fopen(path, "rb");
     READ<uint32_t>(void_, file); // version
     //std::cout << void_ << std::endl;
```
Comment (MrAwesomeRocks) [inline src/Terrain/terrain.cpp:911]: We should set up logging sometime

File: src/Terrain/terrain.cpp [hunk 2]
```diff
@@ -995,6 +968,7 @@ int Terrain::qb_read(const char * path, const std::map<uint32_t, std::pair<const
     }
     fclose(file);
     std::cout << "    tiles read: " << tiles_read << std::endl;
+    free(name);
```
Comment (MrAwesomeRocks) [inline src/Terrain/terrain.cpp:971]: NO NO NO NO NO

NEVER NEVER use malloc and free in C++

USE NEW AND DELETE
OR JUST DO THE RIGHT THING AND USE `std::string` WHICH DOES THE RAII FOR YOU

File: src/Terrain/terrain.hpp [hunk 3]
```diff
@@ -1,6 +1,6 @@
 #ifndef __TERRAIN_HPP__
 #define __TERRAIN_HPP__
-
+// TODO add comments
```
Comment (MrAwesomeRocks) [inline src/Terrain/terrain.hpp:3]: NO TODOs

File: src/Terrain/terrain_mesh.hpp [hunk 4]
```diff
@@ -0,0 +1,104 @@
+//! This is a test. The content tent of this class should be transferred into chunk
```
Comment (MrAwesomeRocks) [inline src/Terrain/terrain_mesh.hpp:1]: THEN MOVE TI

File: src/Terrain/terrain_mesh.hpp [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+// Include GLM
+#include <glm/glm.hpp>
+
+#include "../GUI/meshloader.hpp"
+
+#ifndef __TERRAIN_MESH_HPP__
+#define __TERRAIN_MESH_HPP__
+
+class TerrainMesh : public MeshLoader::SingleComplexMesh {
```
Comment (MrAwesomeRocks) [inline src/Terrain/terrain_mesh.hpp:19]: Functions need to be moved to a cpp file

File: src/Terrain/terrain_mesh.hpp [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    TerrainMesh& operator=(const TerrainMesh& obj){
+        vertex_buffer_ = obj.vertex_buffer_;
+        color_buffer_ = obj.color_buffer_;
+        normal_buffer_ = obj.normal_buffer_;
+        element_buffer_ = obj.element_buffer_;
+        num_vertices_ = obj.num_vertices_;
+        return *this;
+    }
+
+    TerrainMesh(){};
```
Comment (MrAwesomeRocks) [inline src/Terrain/terrain_mesh.hpp:43]: Do you really want this?

File: src/Terrain/terrain_mesh.hpp [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+    TerrainMesh(std::vector<unsigned short> &indices,
```
Comment (MrAwesomeRocks) [inline src/Terrain/terrain_mesh.hpp:45]: How do you make sure this doesn't run multiple times.

File: src/Terrain/terrain_mesh.hpp [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+// Include GLFW
+#include <GLFW/glfw3.h>
+
+// Include GLM
+#include <glm/glm.hpp>
+
+#include "../GUI/meshloader.hpp"
+
+#ifndef __TERRAIN_MESH_HPP__
+#define __TERRAIN_MESH_HPP__
```
Comment (MrAwesomeRocks) [inline src/Terrain/terrain_mesh.hpp:17]: Need path or pragma once

File: src/Terrain/terrain_mesh.hpp [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        return normal_buffer_;
+    }
+
+    GLuint get_vertex_buffer() const override {
+        return vertex_buffer_;
+    }
+
+    unsigned int get_num_vertices() const override {
+        return num_vertices_;
+    }
```
Comment (MrAwesomeRocks) [inline src/Terrain/terrain_mesh.hpp:101]: Should be marked inline

File: src/Terrain/terrain_mesh.hpp [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        num_vertices_ = obj.num_vertices_;
+    };
+    TerrainMesh& operator=(const TerrainMesh& obj){
+        vertex_buffer_ = obj.vertex_buffer_;
+        color_buffer_ = obj.color_buffer_;
+        normal_buffer_ = obj.normal_buffer_;
+        element_buffer_ = obj.element_buffer_;
+        num_vertices_ = obj.num_vertices_;
+        return *this;
+    }
```
Comment (MrAwesomeRocks) [inline src/Terrain/terrain_mesh.hpp:41]: These are weird comment which operator this is

File: src/Terrain/tile.hpp [hunk 11]
```diff
@@ -1,13 +1,13 @@
 #ifndef __TILE_HPP__
 #define __TILE_HPP__
-
+// TODO comments
```
Comment (MrAwesomeRocks) [inline src/Terrain/tile.hpp:3]: NO TODOs

File: src/Terrain/tile.hpp [hunk 12]
```diff
@@ -18,9 +18,9 @@ struct TilePCompare{
 
 class Tile {  // this represents a tile in the world
 private:
-    int x; // The x index
-    int y; // The y index
-    int z; // The z index
+    uint16_t x; // The x index
+    uint16_t y; // The y index
+    uint16_t z; // The z index
```
Comment (MrAwesomeRocks) [inline src/Terrain/tile.hpp:23]: ints are 32 bits on x86_64

File: src/Terrain/tile.hpp [hunk 13]
```diff
@@ -66,12 +66,12 @@ class Tile {  // this represents a tile in the world
...
-    void add_adjacent(Tile *tile, OnePath type);
-    void add_adjacent(std::map<Tile *, OnePath>::iterator it, Tile *tile, OnePath type);
+    void add_adjacent(Tile *tile, UnitPath type);
+    void add_adjacent(std::map<Tile *, UnitPath>::iterator it, Tile *tile, UnitPath type);
     void clear_adjacent();
 
-    std::map<Tile *,OnePath, TilePCompare> & get_adjacent_map() { return adjacent; };
-    const std::map<Tile *,OnePath, TilePCompare> & get_adjacent_map() const { return adjacent; };
+    std::map<Tile *,UnitPath, TilePCompare> & get_adjacent_map() { return adjacent; };
+    const std::map<Tile *,UnitPath, TilePCompare> & get_adjacent_map() const { return adjacent; };
```
Comment (MrAwesomeRocks) [inline src/Terrain/tile.hpp:74]:
```suggestion
    inline std::map<Tile *,UnitPath, TilePCompare> & get_adjacent_map() { return adjacent; };
    inline const std::map<Tile *,UnitPath, TilePCompare> & get_adjacent_map() const { return adjacent; };
```

File: src/Terrain/unitpath.hpp [hunk 14]
```diff
@@ -1,8 +1,8 @@
-#ifndef __ONEPATH_HPP__
-#define __ONEPATH_HPP__
+#ifndef __UNITPATH_HPP__
+#define __UNITPATH_HPP__
```
Comment (MrAwesomeRocks) [inline src/Terrain/unitpath.hpp:2]: Need path or pragma once

File: src/Terrain/unitpath.hpp [hunk 15]
```diff
@@ -1,8 +1,8 @@
...
 
 #include <cstdint>
-
+// TODO add comments
```
Comment (MrAwesomeRocks) [inline src/Terrain/unitpath.hpp:5]: NO TODOs

File: src/Util/voxelutility.cpp [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+
+    //This is partially from goxel with GPL license
+    std::cout << "Reading form " << path << "\n";
+    FILE *file;
+    //int x, y, z; // position in terrain
+    uint32_t void_, compression; // void int 32 used to read 32 bites without saving it.
+    int32_t x_center, y_center, z_center;
+    uint8_t v[4];
+
+    file = fopen(path, "rb");
```
Comment (MrAwesomeRocks) [inline src/Util/voxelutility.cpp:20]: File IO should be done the RAII way: https://cplusplus.com/doc/tutorial/files/

File: src/Util/voxelutility.cpp [hunk 17]
```diff
@@ -0,0 +1,78 @@
+#include <fstream>
```
Comment (MrAwesomeRocks) [inline src/Util/voxelutility.cpp:1]: This whole file is duplicated form terrain.cpp

File: src/Util/voxelutility.cpp [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+    READ<uint32_t>(void_, file);   // vmask
+    //std::cout << void_ << std::endl;(void) void_;
+    READ<uint32_t>(void_, file);
+    //std::cout << void_ << std::endl;(void) void_;
+    // none of these are used
+
+    int8_t name_len;
+    READ<int8_t>(name_len, file);
+    std::cout << "name length: " << (int) name_len << std::endl;
+    char* name = (char*) malloc (name_len);
```
Comment (MrAwesomeRocks) [inline src/Util/voxelutility.cpp:38]: NEVER USE MALLOC/FREE IN C++ USE NEW/DELETE
AND NEVER USE NEW/FREE WHEN YOU CAN USE THE STL

USE `std::string`

File: src/Util/voxelutility.cpp [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+    READ<int32_t>(x_center, file); // x
+    READ<int32_t>(z_center, file); // z
+    READ<int32_t>(y_center, file); // y
+    center = {x_center, y_center, z_center};
+    std::cout << "    max X: " << X_max << std::endl;
+    std::cout << "    max Y: " << Y_max << std::endl;
+    std::cout << "    max Z: " << Z_max << std::endl;
+    data.resize(X_max * Y_max * Z_max);
+    int tiles_read=0;
+    if (static_cast<bool>(compression)){
```
Comment (MrAwesomeRocks) [inline src/Util/voxelutility.cpp:56]:
```suggestion
    if (compression){
```
true is not 0

File: src/Util/voxelutility.cpp [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+        std::cout << "There is compression!" << std::endl;
```
Comment (MrAwesomeRocks) [inline src/Util/voxelutility.cpp:57]: And throw an error?

File: src/Util/voxelutility.hpp [hunk 21]
```diff
@@ -0,0 +1,32 @@
+#ifndef __VOXEL_UTILITY_HPP__
+#define __VOXEL_UTILITY_HPP__
```
Comment (MrAwesomeRocks) [inline src/Util/voxelutility.hpp:2]: Add path or pragma once

File: src/Util/voxelutility.hpp [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+// TODO comments
+#include <cstdint>
+#include <iostream>
+#include <fstream>
+#include <vector>
+#include <cstdint>
```
Comment (MrAwesomeRocks) [inline src/Util/voxelutility.hpp:8]: This is double included

File: src/Util/voxelutility.hpp [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+     std::vector<int>& size);
+
+    template <typename T>
+    void WRITE(T v, FILE * file){
+        fwrite(&v, sizeof(T), 1, file);
+    }
+    template <typename T>
+    void READ(T &v, FILE * file){
+        fread(&v, sizeof(T), 1, file);
+    }
```
Comment (MrAwesomeRocks) [inline src/Util/voxelutility.hpp:24]: Oh I hate this.

Streams exist

File: src/Util/voxelutility.hpp [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+
+    inline uint32_t compress_color(uint8_t v[4]){
+        return (uint32_t)v[3] | (uint32_t)v[2] << 8 | (uint32_t)v[1] << 16 | (uint32_t)v[0] << 24;
+    }
```
Comment (MrAwesomeRocks) [inline src/Util/voxelutility.hpp:28]: Explain how the compression works

and where is the decompression

File: src/main.cpp [hunk 25]
```diff
@@ -138,437 +141,39 @@ int path_finder_test(const char *path, const char *save_path)
     return 0;
 }
 
-int GUITest(const char *path)
-{
-    World world(path);
+void get_mesh(const char * path,
```
Comment (MrAwesomeRocks) [inline src/main.cpp:144]: std::string

File: src/main.cpp [hunk 26]
```diff
@@ -138,437 +141,39 @@ int path_finder_test(const char *path, const char *save_path)
...
-    glewExperimental = true; // Needed for core profile
-    if (glewInit() != GLEW_OK) {
-        fprintf(stderr, "Failed to initialize GLEW\n");
-        getchar();
-        glfwTerminate();
-        return -1;
+    world.get_mesh_greedy(indices,
+            indexed_vertices,
+            indexed_colors,
+            indexed_normals);
```
Comment (MrAwesomeRocks) [inline src/main.cpp:157]: Isn't this for rendering?

File: src/main.cpp [hunk 27]
```diff
@@ -138,437 +141,39 @@ int path_finder_test(const char *path, const char *save_path)
...
 }
 
 int main(int argc, char **argv)
 {
     if (argc == 1) {
-        return path_finder_test("../SavedTerrain/pathfinder_input.qb",
-                                "../SavedTerrain/pathfinder_output.qb");
+        //return path_finder_test("../SavedTerrain/pathfinder_input.qb",
+        //                        "../SavedTerrain/pathfinder_output.qb");
+        return GUITest("../data/Models/DefaultTree.qb");
```
Comment (MrAwesomeRocks) [inline src/main.cpp:176]: NO HARD CODING

File: src/main.cpp [hunk 28]
```diff
@@ -578,6 +183,8 @@ int main(int argc, char **argv)
         } else {
             save_all_terrain(biome_data);
         }
+    } else if (std::string(argv[1]) == "--GenerateTerrain") {
+        return test1(argv[2]);
```
Comment (MrAwesomeRocks) [inline src/main.cpp:187]: This function needs a better name

**Review by MrAwesomeRocks — CHANGES_REQUESTED** [review]
Oh boy
---
* `mesh.cpp` has HORRIBLE variable names, everything is one letter. Please name more descriptively.
* I trust you on the shaders
* You should move the shaders to a `resources` directory at repo root
* I feel like there's some code duplication, like you have multiple files loading shaders and rendering things.
* NEVER USE MALLOC. You shouldn't even use new! C++ comes with RAII in the STL that does everything for you.
* **USE `std::string`!!!!!!!**
* Look into [`#pragma once`](https://en.wikipedia.org/wiki/Pragma_once) or add paths to your include guards
* Do file IO with streams: https://cplusplus.com/doc/tutorial/files/
* `voxelutility` is really I/O Utility and also duplicated in terrain.cpp

File: CMakeLists.txt [hunk 29]
```diff
@@ -29,7 +29,11 @@ add_executable(Fun_Game ${SOURCES1} ${SOURCES2} ${SOURCES3})
 target_link_libraries(Fun_Game glfw)
 target_link_libraries(Fun_Game OpenGL::GL)
 target_link_libraries(Fun_Game GLEW::GLEW)
-target_link_libraries(Fun_Game glm::glm)
+if (WIN32)
+  target_link_libraries(Fun_Game glm)
+else()
+  target_link_libraries(Fun_Game glm::glm)
+endif()
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply CMakeLists.txt:36]: it works?

File: src/Entity/mesh.hpp [hunk 30]
```diff
@@ -0,0 +1,59 @@
...
+    std::vector<int> center_;
+
+    std::vector<std::uint16_t> indices_;
+    std::vector<glm::vec3> indexed_vertices_;
+    std::vector<glm::vec3> indexed_colors_;
+    std::vector<glm::vec3> indexed_normals_;
+
+    int get_position(int x, int y, int z) const ;
+
+    void generate_mesh(std::vector<uint32_t> data,
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/Entity/mesh.hpp:54]: learn to read

File: src/Entity/mesh.hpp [hunk 31]
```diff
@@ -0,0 +1,59 @@
...
+ */
+class Mesh{
+public:
+    Mesh(const char* path);
+    void get_mesh(std::vector<unsigned short> &indices,
+                  std::vector<glm::vec3> &indexed_vertices,
+                  std::vector<glm::vec3> &indexed_colors,
+                  std::vector<glm::vec3> &indexed_normals) const;
+
+    void load_from_qb(const char * path);
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/Entity/mesh.hpp:40]: it is now private

File: src/Entity/mesh.hpp (hunk 30 above)
[the remark this answers is not in the document]
Comment (MrAwesomeRocks) [reply src/Entity/mesh.hpp:54]: WDYM

Private ends with an _

File: src/Entity/mesh.cpp [hunk 32]
```diff
@@ -0,0 +1,198 @@
...
+        // Compute mask
+        q[axis] = 1;
+        for (x[axis] = -1; x[axis] < dims[axis];) {
+            // printf("x: %d, %d, %d\n", x[0], x[1], x[2]);
+            // printf("q: %d, %d, %d\n", q[0], q[1], q[2]);
+
+            std::size_t counter = 0;
+            for (x[v] = 0; x[v] < dims[v]; ++x[v])
+                for (x[u] = 0; x[u] < dims[u]; ++x[u], ++counter) {
+                    const uint32_t a = 0 <= x[axis]
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/Entity/mesh.cpp:75]: ok?

File: src/GUI/renderer.cpp [hunk 33]
```diff
@@ -0,0 +1,258 @@
...
+        depth_projection_matrix_ * depth_view_matrix_ * depthModelMatrix;
+
+    // Compute the MVP matrix from keyboard and mouse input
+    controls::computeMatricesFromInputs(window);
+    glm::mat4 projection_matrix = controls::get_projection_matrix();
+    glm::mat4 view_matrix = controls::get_view_matrix();
+    // glm::mat4 ModelMatrix = glm::mat4(1.0);
+    glm::mat4 MVP = projection_matrix * view_matrix;  // * ModelMatrix;
+
+    glm::mat4 biasMatrix(0.5, 0.0, 0.0, 0.0, 0.0, 0.5, 0.0, 0.0, 0.0, 0.0, 0.5,
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/GUI/renderer.cpp:112]: ya I don't know

File: src/GUI/renderer.cpp [hunk 34]
```diff
@@ -0,0 +1,258 @@
...
+
+    glUniform3f(light_direction_ID_, light_direction_.x, light_direction_.y,
+                light_direction_.z);
+
+    // Bind our texture in Texture Unit 1
+    glActiveTexture(GL_TEXTURE1);
+    glBindTexture(GL_TEXTURE_2D, depth_texture_);
+    glUniform1i(shadow_map_ID_, 1);
+
+    for (std::shared_ptr<MeshLoader::SingleComplexMesh> mesh :
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/GUI/renderer.cpp:131]: render each mesh.
This a allows me to added many meshes with the same shader.

File: src/GUI/renderer.cpp [hunk 35]
```diff
@@ -0,0 +1,258 @@
...
+    glUniformMatrix4fv(depth_bias_ID_multi_, 1, GL_FALSE, &depthBiasMVP[0][0]);
+    glUniform3f(light_direction_ID_multi_, light_direction_.x,
+                light_direction_.y, light_direction_.z);
+
+    // Bind our texture in Texture Unit 0
+    glActiveTexture(GL_TEXTURE1);
+    glBindTexture(GL_TEXTURE_2D, depth_texture_);
+    glUniform1i(shadow_map_ID_multi_, 1);
+
+    for (std::shared_ptr<MeshLoader::MultiComplexMesh> mesh : multis_meshes_) {
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/GUI/renderer.cpp:197]: same

File: src/GUI/renderer.cpp (hunk 34 above)
[the remark this answers is not in the document]
Comment (MrAwesomeRocks) [reply src/GUI/renderer.cpp:131]: Then add a comment

File: src/GUI/renderer.cpp (hunk 35 above)
[the remark this answers is not in the document]
Comment (MrAwesomeRocks) [reply src/GUI/renderer.cpp:197]: Then add a comment

File: src/GUI/renderer.cpp [hunk 36]
```diff
@@ -0,0 +1,258 @@
...
+        glVertexAttribPointer(2,         // attribute
+                              3,         // size
+                              GL_FLOAT,  // type
+                              GL_FALSE,  // normalized?
+                              0,         // stride
+                              (void *)0  // array buffer offset
+        );
+
+        // 4th attribute buffer : transform
+        glEnableVertexAttribArray(3);
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/GUI/renderer.cpp:232]: zero indexing?

so basicily one of the meshes is "indexed" that means that the same mesh is drawn multipul times.
The extra array buffer is the off set of the entire mesh.

File: src/GUI/controls.hpp [hunk 37]
```diff
@@ -3,8 +3,26 @@
...
+ * @brief Get the view matrix
+ * 
+ * @return glm::mat4 projection of world space to camera space
+ */
 glm::mat4 get_view_matrix();
+
+/**
+ * @brief Get the projection matrix
+ * 
+ * @return glm::mat4 homographic projection depending on the FOV
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/GUI/controls.hpp:24]: the world...

File: src/GUI/meshloader.hpp [hunk 38]
```diff
@@ -0,0 +1,63 @@
+// Include GLEW
+#include <GL/glew.h>
+
+#ifndef __MESH_LOADER_HPP__
+#define __MESH_LOADER_HPP__
+
+namespace MeshLoader{
+
+/**
+ * @brief virtual class for non-indexed mesh
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/GUI/meshloader.hpp:10]: www.google.com has answers to the wonders of the universe

File: src/GUI/renderer.cpp [hunk 39]
```diff
@@ -0,0 +1,258 @@
...
+
+void MainRenderer::add_mesh(
+    std::shared_ptr<MeshLoader::SingleComplexMesh> mesh) {
+    singles_meshes_.push_back(std::move(mesh));
+}
+
+void MainRenderer::add_mesh(
+    std::shared_ptr<MeshLoader::MultiComplexMesh> mesh) {
+    multis_meshes_.push_back(std::move(mesh));
+}
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/GUI/renderer.cpp:66]: yes, but I will not use the benifit untill later

File: src/GUI/renderer.cpp [hunk 40]
```diff
@@ -0,0 +1,258 @@
...
+    matrix_ID_multi_ = glGetUniformLocation(programID_multi_, "MVP");
+    view_matrix_ID_multi_ = glGetUniformLocation(programID_multi_, "V");
+    depth_bias_ID_multi_ = glGetUniformLocation(programID_multi_, "DepthBiasMVP");
+    shadow_map_ID_multi_ = glGetUniformLocation(programID_multi_, "shadowMap");
+    light_direction_ID_multi_ =
+        glGetUniformLocation(programID_multi_, "LightInvDirection_worldspace");
+
+    light_direction_ =
+        glm::normalize(glm::vec3(40.0f, 8.2f, 120.69f))  // direction
+        * 128.0f;                                        // length
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/GUI/renderer.cpp:44]: so your not going to belive this but throught complex selestial mechanism, the direction of light chages.

File: src/GUI/shadow_map.cpp [hunk 41]
```diff
@@ -0,0 +1,214 @@
...
+    }
+}
+
+void ShadowMap::add_mesh(std::shared_ptr<MeshLoader::SingleMesh> mesh) {
+    singles_meshes_.push_back(std::move(mesh));
+}
+
+void ShadowMap::add_mesh(std::shared_ptr<MeshLoader::MultiMesh> mesh) {
+    multi_meshes_.push_back(std::move(mesh));
+}
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/GUI/shadow_map.cpp:77]: ya this is the shaodw map not the renderer they are compleatly different

This one only holds the "shape" of the mesh and is used to render the depth buffer. the renderer uses that depth buffer to add shadows

File: src/GUI/shadow_map.cpp [hunk 42]
```diff
@@ -0,0 +1,214 @@
...
+                       mesh->get_num_vertices(),  // count
+                       GL_UNSIGNED_SHORT,         // type
+                       (void *)0                  // element array buffer offset
+        );
+
+        glDisableVertexAttribArray(0);
+    }
+
+    // Use our shader
+    glUseProgram(programID_multi_);
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/GUI/shadow_map.cpp:153]: yes one is for indexed meshes the other is for non-indexed meshes

File: src/GUI/shadow_map.cpp [hunk 43]
```diff
@@ -0,0 +1,214 @@
...
+    shadow_width = w;
+    shadow_height = h;
+
+    // ---------------------------------------------
+    // Render Shadow to Texture - specific code begins here
+    // ---------------------------------------------
+
+    // The framebuffer, which regroups 0, 1, or more textures, and 0 or 1 depth
+    // buffer.
+    FramebufferName = 0;
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/GUI/shadow_map.cpp:41]: look your guess is as good as mine

File: src/GUI/shadow_map.cpp [hunk 44]
```diff
@@ -0,0 +1,214 @@
...
+#include "shader.hpp"
+#include "shadow_map.hpp"
+
+ShadowMap::ShadowMap(int w, int h) {
+    depthTexture = 0;
+    FramebufferName = 0;
+    programID_ = LoadShaders("../src/GUI/Shaders/DepthRTT.vert",
+                             "../src/GUI/Shaders/DepthRTT.frag");
+    programID_multi_ = LoadShaders("../src/GUI/Shaders/DepthRTTInstanced.vert",
+                                   "../src/GUI/Shaders/DepthRTTInstanced.frag");
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/GUI/shadow_map.cpp:22]: they are all different

File: src/Terrain/chunk.hpp [hunk 45]
```diff
@@ -67,7 +67,7 @@ class Chunk
     Terrain* ter_;
     void delNodeGroup(NodeGroup &NG);
     void mergeNodeGroup(NodeGroup &g1, NodeGroup &g2);
-    int Cx_, Cy_, Cz_; // Chunk position. Incremented by 1 so multiply by chunk::size to get tile position.
+    uint16_t Cx_, Cy_, Cz_; // Chunk position. Incremented by 1 so multiply by chunk::size to get tile position.
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/Terrain/chunk.hpp:70]: ??

File: src/Terrain/static_mesh.hpp [hunk 46]
```diff
@@ -0,0 +1,128 @@
...
+        color_buffer_ = obj.color_buffer_;
+        normal_buffer_ = obj.normal_buffer_;
+        element_buffer_ = obj.element_buffer_;
+        transforms_buffer_ = obj.transforms_buffer_;
+        num_vertices_ = obj.num_vertices_;
+        num_models_ = obj.num_models_;
+        return *this;
+    }
+
+    StaticMesh(){};
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/Terrain/static_mesh.hpp:49]: Am I sure of anything?

File: src/Entity/mesh.cpp [hunk 47]
```diff
@@ -0,0 +1,198 @@
...
+                    indexed_colors_.push_back(vector_color);
+                    indexed_colors_.push_back(vector_color);
+                    indexed_colors_.push_back(vector_color);
+                    indexed_colors_.push_back(vector_color);
+
+                    glm::vec3 normal = glm::normalize(
+                                           glm::cross(indexed_vertices_[vertexSize]
+                                                        - indexed_vertices_[vertexSize+1],
+                                                      indexed_vertices_[vertexSize]
+                                                        - indexed_vertices_[vertexSize+2]));
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/Entity/mesh.cpp:159]: your mom

File: src/Entity/mesh.cpp [hunk 48]
```diff
@@ -0,0 +1,198 @@
...
+            for (x[v] = 0; x[v] < dims[v]; ++x[v])
+                for (x[u] = 0; x[u] < dims[u]; ++x[u], ++counter) {
+                    const uint32_t a = 0 <= x[axis]
+                                ? data[get_position(x[0], x[1], x[2])]
+                                : 0;
+                    const uint32_t b = x[axis] < dims[axis] - 1
+                                ? data[get_position(x[0] + q[0], x[1] + q[1], x[2] + q[2])]
+                                : 0;
+                    const bool ba = static_cast<bool>(a);
+                    if (ba == static_cast<bool>(b)) // if both solid or both not solid
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/Entity/mesh.cpp:82]: these are both vars. what are you talking about?

File: src/Entity/mesh.cpp [hunk 49]
```diff
@@ -0,0 +1,198 @@
...
+
+void Mesh::get_mesh(std::vector<unsigned short> &indices,
+                  std::vector<glm::vec3> &indexed_vertices,
+                  std::vector<glm::vec3> &indexed_colors,
+                  std::vector<glm::vec3> &indexed_normals) const {
+    indices = indices_;
+    indexed_vertices = indexed_vertices_;
+    indexed_colors = indexed_colors_;
+    indexed_normals = indexed_normals_;
+
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/Entity/mesh.cpp:196]: ??

File: src/GUI/main_gui.hpp [hunk 50]
```diff
@@ -0,0 +1,225 @@
...
+        // Use our shader
+        glUseProgram(quad_programID);
+
+        // Bind our texture in Texture Unit 0
+        glActiveTexture(GL_TEXTURE0);
+        glBindTexture(GL_TEXTURE_2D, SM.get_depth_texture());
+        // Set our "renderedTexture" sampler to use Texture Unit 0
+        glUniform1i(texID, 0);
+
+        // 1rst attribute buffer : vertices
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/GUI/main_gui.hpp:187]: ?

File: src/Terrain/terrain_mesh.hpp (hunk 6 above)
Comment (AlemSnyder) [reply src/Terrain/terrain_mesh.hpp:43]: I think for shared pointer

File: src/Terrain/static_mesh.hpp [hunk 51]
```diff
@@ -0,0 +1,128 @@
+//! This is a test. The content tent of this class should be transferred into chunk
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/Terrain/static_mesh.hpp:1]: I was wrong, just terrain mesh

File: src/Terrain/static_mesh.hpp [hunk 52]
```diff
@@ -0,0 +1,128 @@
...
+        return *this;
+    }
+
+    StaticMesh(){};
+
+    StaticMesh(std::vector<unsigned short> &indices,
+                std::vector<glm::vec3> &indexed_vertices,
+                std::vector<glm::vec3> &indexed_colors,
+                std::vector<glm::vec3> &indexed_normals,
+                std::vector<glm::vec3> &model_transforms){
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/Terrain/static_mesh.hpp:55]: The problem was that I was sending my pointer to the shadow relm. there should be nothing wrong with renning this twice

File: src/Terrain/terrain_mesh.hpp (hunk 7 above)
Comment (AlemSnyder) [reply src/Terrain/terrain_mesh.hpp:46]: that's not bad

File: src/Terrain/terrain_mesh.hpp (hunk 4 above)
Comment (AlemSnyder) [reply src/Terrain/terrain_mesh.hpp:1]: different pull request

File: src/GUI/main_gui.hpp [hunk 53]
```diff
@@ -0,0 +1,225 @@
...
+    std::vector<glm::vec3> indexed_normals;
+
+    world.get_mesh_greedy(indices, indexed_vertices, indexed_colors,
+                          indexed_normals);
+
+    std::vector<std::uint16_t> indices_tree;
+    std::vector<glm::vec3> indexed_vertices_tree;
+    std::vector<glm::vec3> indexed_colors_tree;
+    std::vector<glm::vec3> indexed_normals_tree;
+    Mesh mesh("../data/Models/DefaultTree.qb");
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/GUI/main_gui.hpp:38]: This pull request is about a different file, and that rediced hard coding.

[the remark this answers is not in the document]
Comment (MrAwesomeRocks) [reply src/GUI/main_gui.hpp:38]: Then undo the changes.

**Every** change, even if not explicitly part of the main PR change, is reviewed.

File: src/Terrain/terrain_mesh.hpp (hunk 4 above)
Comment (MrAwesomeRocks) [reply src/Terrain/terrain_mesh.hpp:1]: Then undo the changes.

**Every** change, even if not explicitly part of the main PR change, is reviewed.

**Comment by AlemSnyder** [comment]
I did everything I could reasonably do in the PR. I will make another that refactors things, and adds comments to terrain.hpp/tile/node if you rewrite the IO.

**Pull request merged**